src/extra/gfx/shader/translator.cpp

```diff
@@ -0,0 +1,230 @@
+  using Callback = std::function<void(TShard *shard, TranslationContext &)>;
+
+  Callback callback;
+
+  TranslateWrapperExternal(Callback callback) : callback(callback){};
+  void translate(ShardPtr shard, TranslationContext &context) {
+    ShardWrapper *wrapper = reinterpret_cast<ShardWrapper *>(shard);
+    TShard *innerShard = &wrapper->shard;
+
+    callback(innerShard, context);
+  }
+};
+
+// Translator is called as instance shard.translate(context&)
+#define REGISTER_SHADER_SHARD(_name, _class)                    \
+  {                                                             \
+    static TranslateWrapper<_class> instance{};                 \
+    getTranslationRegistry().registerHandler(_name, &instance); \
+    REGISTER_SHARD(_name, _class);                              \
+  }
+
+// Translator is called as static TranslatorClass::translate(ShardClass*, context&)
+#define REGISTER_EXTERNAL_SHADER_SHARD(_translatorClass, _name, _class)             \
+  {                                                                                 \
+    static TranslateWrapperExternal<_class> instance(&_translatorClass::translate); \
+    getTranslationRegistry().registerHandler(_name, &instance);                     \
+  }
+
+// Translator is called as static TranslatorClass<TShard>::translate(TShard*, context&)
+#define REGISTER_EXTERNAL_SHADER_SHARD_T1(_translatorClass, _name, _class)                  \
+  {                                                                                         \
+    static TranslateWrapperExternal<_class> instance(&_translatorClass<_class>::translate); \
+    getTranslationRegistry().registerHandler(_name, &instance);                             \
+  }
+
+// Translator is called as static TranslatorClass<TShard, T1>::translate(TShard*, context&)
+#define REGISTER_EXTERNAL_SHADER_SHARD_T2(_translatorClass, _name, _class, _t1)                  \
+  {                                                                                              \
+    static TranslateWrapperExternal<_class> instance(&_translatorClass<_class, _t1>::translate); \
+    getTranslationRegistry().registerHandler(_name, &instance);                                  \
+  }
+
+namespace LinAlg {
+using namespace shards::Math::LinAlg;
+}
+template <SHType ToType> using ToNumber = shards::ToNumber<ToType>;
+
+void registerTranslatorShards() {
+  REGISTER_SHARD("GFX.TestTranslator", TestTranslator);
+
+  // Literal copy-paste into shader code
+  REGISTER_SHADER_SHARD("Shader.Literal", Literal);
+  REGISTER_SHADER_SHARD("Shader.ReadInput", gfx::shader::Read<blocks::ReadInput>);
+  REGISTER_SHADER_SHARD("Shader.ReadBuffer", gfx::shader::ReadBuffer);
+  REGISTER_SHADER_SHARD("Shader.ReadGlobal", gfx::shader::Read<blocks::ReadGlobal>);
+  REGISTER_SHADER_SHARD("Shader.WriteGlobal", gfx::shader::Write<blocks::WriteGlobal>);
+  REGISTER_SHADER_SHARD("Shader.WriteOutput", gfx::shader::Write<blocks::WriteOutput>);
```

Comment:
Do we need `Shader.` prefix I wonder 🤔

Reply:
I wanted to avoid it using the context mechanism (#167) where they would only be available inside shader wires.
But for now I think it's better to avoid the name collisions like this.